src/replica/HttpAsyncClient.h

```diff
@@ -0,0 +1,312 @@
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+// Third-party headers
+#include "boost/asio.hpp"
+#include "boost/beast.hpp"
+
+// Qserv headers
+#include "util/Mutex.h"
+
+// This header declarations
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+ * @brief Class HttpAsyncClient represents a simple asynchronous interface for
+ * communicating over the HTTP protocol.
+ * 
+ * The implementation of the class invokes a user-supplied callback (lambda) function
+ * upon a completion or a failure of the request.
+ * 
+ * Here is an example of using the class to pull a file and dump its content on
+ * to the standard output stream:
+ * @code
+ *   boost::asio::io_service io_service;
+ *   std::shared_ptr<HttpAsyncClient> const reader =
+ *       HttpAsyncClient::create(
+ *           [](auto const& reader) {
+ *               std::cout << "state: " << reader->state() << std::endl;
+ *           },
+ *           io_service,
+ *           "GET", "http://my.host.domain/data/chunk_0.txt");
+ *   });
+ *   reader->start();
+ *   io_service.run();
+ * @code
+ * 
+ * @note Once the start() method gets called w/o throwing an exception, the algorithm
+ *   can make repeated attempts to establish a connection to the Web server even if
+ *   the server's host and/or port won't be resolved, or if the connection attempts
+ *   will be failing, or if attempts to send a request will be failing. A short delay
+ *   will be introduced after each unsuccessfull attempt. The request will go into
+ *   the FAILED state if 1) there will be a failure in receiving server's response,
+ *   2) in case of incorrect data received from the server, or 3) the result is too
+ *   large.
+ *
+ * @note The implementation will open and close a new connection for each request.
+ * @note The implementation doesn't support TLS/SSL-based HTTPS protocol.
+ */
+class HttpAsyncClient: public std::enable_shared_from_this<HttpAsyncClient> {
+public:
+    /// The function type for notifications on the completion of the operation.
+    typedef std::function<void(std::shared_ptr<HttpAsyncClient> const&)> CallbackType;
+
+    enum class State: int {
+        CREATED = 0,        ///< The object was created and no request was sent
+        IN_PROGRESS,        ///< The request is still in progress
+        FINISHED,           ///< Final state: the operation succeeded, result is available 
+        FAILED,             ///< Final state: the operation failed
+        BODY_LIMIT_ERROR,   ///< Final state: the operation failed because the response's body is larger than requested
+        CANCELLED,          ///< Final state: the operation was explicitly cancelled after it was started
+        EXPIRED             ///< Final state: the operation was aborted due to the timeout expiration
+    };
+
+    /// @return The string representation of the state.
+    /// @throw std::invalid_argument For unknown values of the input parameter.
```

Comment:
This often happens (I have seen this many times) if there is silent memory corruption. Tracking the out-of-bound input values is very helpful in the early detection of problems.

Reply:
That's an odd argument. If you have a memory corruption I think you cannot even reason about what your program does.

Reply:
At least, I will see the problem earlier. This helped in the past. The out-of-bound enumerator values retrieved from the data members of some class usually indicate that an object of that class got destroyed, corrupt, or never created. This is never a problem in Python because Python doesn't know anything about such things as "threads" or "event-driven" programming models 😄 
What I could do here is to remove the notion of the exception from the class contract while still leaving this extra check in the body of the method. This may prevent confusion.